### PR TITLE
refs #66. Added a demo page to show how to use the API (folder `example`...

### PR DESCRIPTION
**Pull request by schettino72**
...).

### CODE REVIEW

File: example/src/main/scala/example/Example.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  // name of example
+  val title: String
+  // supporting HTML to run the example
+  val html: String
+  // scala.js code to be executed
+  def run: Unit
+}
+
+
+object Alert extends Example {
```
Comment (lihaoyi) [inline example/src/main/scala/example/Example.scala]: Could we add a link to the readme, so people can easily get to this code to look at it? I'd guess people would want to see the code more than they'd want to see the working demo.

Comment (schettino72) [inline example/src/main/scala/example/Example.scala]: Done.

Actually I want to include the scala source into the demo page itself. But I didnt figure out how to print the source of a function in scala.

File: README.md [hunk 2]
```diff
@@ -38,6 +38,8 @@ then enjoy the types available in `org.scalajs.dom`.
 scalajs-dom 0.6 is built and published for Scala.js 0.5.0 and following in
 the 0.5.x series, with both Scala 2.10 and 2.11.
 
+The project include a demo page with some examples [(source)](https://github.com/scala-js/scala-js-dom/blob/master/example/src/main/scala/example/Example.scala). Compile it with `sbt example/compile:fastOptJS` then open the file `example/index.html` in your browser.
```
Comment (sjrd) [inline README.md]: * The include<ins>s</ins>
* Put the () of "(sources)" outside the link. In other words, only "source" should be the link, not the () around it.
* `sbt example/fastOptJS` (no need for `compile:` here).

File: build.sbt [hunk 3]
```diff
@@ -55,3 +55,11 @@ pomExtra := (
 )
 
 pomIncludeRepository := { _ => false }
+
+
+
+lazy val example = project.in(file("example"))
```
Comment (sjrd) [inline build.sbt]: One blank line only. We never use more than 1 blank line in a row.

File: build.sbt [hunk 4]
```diff
@@ -55,3 +55,11 @@ pomExtra := (
...
+  .dependsOn(root)
+  .enablePlugins(ScalaJSPlugin)
+  .settings(scalaVersion := "2.11.2")
+
```
Comment (sjrd) [inline build.sbt]: No empty line at the end of the file. There should be a final EOL, but not *2* EOLs.

File: example/index.html [hunk 5]
```diff
@@ -0,0 +1,47 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="UTF-8">
+    <title>The Scala.js DOM demo</title>
+    <style>
```
Comment (sjrd) [inline example/index.html]: Please put the stylesheet in its own .css file.

File: example/src/main/scala/example/Example.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package example
+
+import scala.scalajs.js.JSApp
+
+import org.scalajs.dom
+
+
```
Comment (sjrd) [inline example/src/main/scala/example/Example.scala]: Same thing about blank lines: never more than 1. There are a few more occurrences in this file.

File: example/src/main/scala/example/Example.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+trait Example {
+  // name of example
```
Comment (sjrd) [inline example/src/main/scala/example/Example.scala]: name of <ins>the</ins> example

File: example/src/main/scala/example/Example.scala [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  val title: String
+  // supporting HTML to run the example
+  val html: String
+  // scala.js code to be executed
+  def run: Unit
```
Comment (sjrd) [inline example/src/main/scala/example/Example.scala]: As a side-effecting method, this should be defined with an empty pair of parentheses `()`.

File: example/src/main/scala/example/Example.scala [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  * Right side #output contains the resulting tree of the example execution.
+  * TODO: it would be nice to show the original scala code
+  */
+object DemoApp extends JSApp {
+  val examples = dom.document.getElementById("examples")
+  val title = dom.document.getElementById("ex-title")
+  val sandbox = dom.document.getElementById("ex-sandbox")
+
+  // include given Example in the list of examples
+  def addExample(ex: Example){
```
Comment (sjrd) [inline example/src/main/scala/example/Example.scala]: Don't use the old procedural notation. Write `: Unit = {` explicitly.

File: example/src/main/scala/example/Example.scala [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    link.href = "#" + ex.title
+    link.textContent = ex.title
+    link.onclick = (_: dom.MouseEvent) => showExample(ex)
+
+    div.appendChild(link)
+    examples.appendChild(div)
+  }
+
+  // callback to update UI and run given example
+  def showExample(ex: Example){
```
Comment (sjrd) [inline example/src/main/scala/example/Example.scala]: `: Unit = {`

File: example/src/main/scala/example/Example.scala [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    title.textContent = ex.title
+    sandbox.innerHTML = ex.html
+    ex.run
```
Comment (sjrd) [inline example/src/main/scala/example/Example.scala]: `()`

File: example/src/main/scala/example/Example.scala [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+  val html: String
+  // scala.js code to be executed
+  def run: Unit
+}
+
+
+object Alert extends Example {
+  val title: String = "alert"
+  val html = ""
+  def run: Unit = {
```
Comment (sjrd) [inline example/src/main/scala/example/Example.scala]: `()`

File: example/src/main/scala/example/Example.scala [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    dom.alert("Hi from Scala-js-dom")
+  }
+}
+
+
+object NodeAppendChild extends Example {
+  val title = "Node.appendChild"
+  val html = """<div id="container"><div>"""
+  def run: Unit = {
```
Comment (sjrd) [inline example/src/main/scala/example/Example.scala]: `()`

**Comment by sjrd** [comment]
The commits should be squashed, and the commit message should be:

> Fix #66: Add a demo page to show how to use the API (folder `example`).

to follow commit message conventions, i.e.,

* "Fix #x: Message." when fixing an issue
* Use imperative present tense instead of past tense (Add instead of Added)

**Comment by lihaoyi** [comment]
I have a merged-&-updated version of this that includes all the examples into the readme inline. Will put it up when I have some time

**Comment by lihaoyi** [comment]
For some reason it's not turning up on github pages (maybe just need to wait a bit longer?) but here's what I have so far

https://dl.dropboxusercontent.com/u/7997532/output/index.html

What do you think? I took @schettino72's two examples and added a few more that I thought would be interesting. Not comprehensive, but pretty!

**Comment by lihaoyi** [comment]
Here it is on gh-pages

http://scala-js.github.io/scala-js-dom/

**Comment by schettino72** [comment]
@lihaoyi I guess you forgot to push the source code.

It looks great! 

Though originally I intend this to be used to quickly try some code without the need to setup another project...

**Comment by lihaoyi** [comment]
Yeah I haven't pushed the source code. Your use case differs from what I imagined; I thought this would be for people to look at online and see examples =P

**Comment by lihaoyi** [comment]
The source code for the index page is now at https://github.com/scala-js/scala-js-dom/blob/scalatex/readme/src/main/scalatex/readme/Index.scalatex, and the rest of that branch contains the remaining changes (e.g. your example project). Build with readme/run.

**Comment by schettino72** [comment]
Just to be clear... this can be closed or you guys expect me to anything else with this?

Thanks for the reviews and great work!

**Pull request closed without merging**